internal/controllers/imagerepository_controller.go

```diff
@@ -516,8 +519,59 @@ func (r *ImageRepositoryReconciler) scan(ctx context.Context, obj *imagev1.Image
 		return 0, err
 	}
 
+	storedTags := make([]database.Tag, 0, len(filteredTags))
+
+	if r.FetchDigests {
```

Comment:
Do we intend to use the tag digests for anything other than storing the latest image digest in the state? If not, I'm wondering if the overhead of making so many registry API calls is worth it. Maybe instead of fetching the digest for all tags stored in the database we only retrieve it for the latest tag instead?

Reply:
Does the registry API has an endpoint for looking up the tags associated with a given digest?

Reply:
Nope, the relationship is unidirectional tag -> digest. The API endpoint is [documented here](https://docs.docker.com/registry/spec/api/#get-manifest) and while it accepts either a tag or a digest, the response for a digest will not include any tags information.

Reply:
> Maybe instead of fetching the digest for all tags stored in the database we only retrieve it for the latest tag instead?

I really like this idea, this would mean adding the digest query to the ImagePolicy reconciler, which would reduce the registry calls by 99%.

Reply:
Can the `latest` tag be configurable as suggested by @relu [here](https://github.com/fluxcd/image-reflector-controller/issues/370)? This way I can make multiple clusters be in sync with multiple different `latest` tags, e.g. `latest-staging`, `latest-prod`, etc.

Reply:
> Can the `latest` tag be configurable as suggested by @relu [here](https://github.com/fluxcd/image-reflector-controller/issues/370#issuecomment-1500871740)? This way I can make multiple clusters be in sync with multiple different `latest` tags, e.g. `latest-staging`, `latest-prod`, etc.

I guess that's what you meant here: `fetch the digest in the ImagePolicy reconciler`

Reply:
> If ImagePolicy automatically periodically checks and updates the digest of the selected tag, I think we can get away without introducing any new policy type and using the existing constructs. The ImagePolicy .spec.filterTags can be used to filter an exact pattern for a specific tag, say "latest". Whenever "latest" tag is overwritten, the ImagePolicy could query the registry in the provided .spec.interval and get the new digest and update the status which can then be used by image automation.

Indeed, this solution would seamlessly solve the problem of fetching the digest for the selected tag for all the three available policies. :+1: 

In my use case I guess the spec would look like this?

```yaml
spec:
  imageRepositoryRef:
    name: app-img-repo
  filterTags:
    pattern: '^latest-staging$'
  interval: 1m
```

I.e. the `.spec.policy` field becomes optional?

Reply:
> It's easier to continue using alphabetical policy which won't have any actual effect on the result.

Yeah it works, but it's also weird/confusing being forced to specify something that won't have any actual effect. In this case I would personally prefer a policy option that clearly states your intention of watching a static tag, say `latest`, whose underlying image/digest is changing over time (maybe start calling it "dynamic tag" instead?).

> This would require making the .spec.imageRepositoryRef optional, such that either ImageRepository reference or a static tag name must be specified in the ImagePolicy.

And how would ImagePolicy know the repo URL if just a static (dynamic?) tag name is specified? When you say tag name here, do you mean a fully-qualified one including the repo URL? I think the point of having a static (dynamic?) tag policy is that the ImagePolicy reconciler would periodically fetch the digest, and then ImageUpdateAutomation could push a git change with the digest, and not just the static (dynamic?) tag (which wouldn't require ImageUpdateAutomation at all). This is useful because it forces workloads to be redeployed when the underlying image pointed to by a tag changes, since the workload configuration is effectively being changed (with the new digest of the fixed tag).

In other words, this use case would still require repository scanning, just not the current scan performed by ImageRepository which is just for listing the tags.

Reply:
I've finally gotten to catching up on this thread and I want to say this escalated quickly 😂. Thanks all for the feedback. Here's my thoughts:

> Can the latest tag be configurable as suggested by @relu https://github.com/fluxcd/image-reflector-controller/issues/370#issuecomment-1500871740? This way I can make multiple clusters be in sync with multiple different latest tags, e.g. latest-staging, latest-prod, etc.

I really want to see this feature land with as little change to the API as possible. That is going to be the premise of this PR going forward. The sugggested change is only very slightly related to this PR so I'd like to keep it out of scope for now.

Another premise I find important to point out is to not encourage usage of static tags, i.e. `latest` and the likes. We always articulated this premise in requests for such a feature and provide proper ways to use tags, e.g. in https://fluxcd.io/flux/guides/sortable-image-tags/. Employing static tags literally opposes the benefits of content-based addressing, i.e. immutability.

The goal I'm envisioning here is to stay as close as possible to the current use cases of ImagePolicy and merely additionally provide a digest of the latest tag that can be used instead of the tag to really pin down an image to the content address which never changes (which basically is a security and reproducibility feature more than anything else).

> ImageRepository operates on tags, it won't be able to identify a change in the underlying content if a tag is overwritten

This is the main challenge to be resolved right now and also a question of API guarantees: If the ImageRepository is reconciled and provides a list of tags to the ImagePolicy reconciler which in turn pulls the digest of the latest tag what is the guarantee the API provides? The simplest thing to say is that there is no guarantee of the latest digest to be provided after it has been reconciled once, i.e. using image digests with static tags such as `latest` will not necessarily result in immediate reflection of the latest digest in the ImagePolicy object.

Reply:
> Another premise I find important to point out is to not encourage usage of static tags, i.e. latest and the likes. We always articulated this premise in requests for such a feature and provide proper ways to use tags, e.g. in https://fluxcd.io/flux/guides/sortable-image-tags/. Employing static tags literally opposes the benefits of content-based addressing, i.e. immutability.

That makes sense. Does this position imply `wontfix` for #370? If so, could you please update the issue? Thanks in advance!